Enhance audio handling by ensuring ambient noise is only transmitted when no content is playing.

abdulrahmanmajid opened the pull request:
Update audio playing status logic to differentiate between content and ambient audio or system audios such as backchanneling.

Improved Azure synthesizer with connection pooling and comprehensive latency tracking for better performance, dropped from avg 0.7 1.0 ms latencies to consistent average 0.2 - 0.4 ms Synthisizer_first_chunk_latency by changing code to align with the new documentation for latency improvements.

abdulrahmanmajid committed:
Enhance audio handling by ensuring ambient noise is only transmitted when no content is playing.

Update audio playing status logic to differentiate between content and ambient audio or system audios such as backchanneling.

massivly Improved Azure synthesizer with connection pooling and comprehensive latency tracking for better performance, droped from avg 0.7 1.0 ms latencies to consistent average 0.2 - 0.4 ms Synthisizer_first_chunk_latency by changing code to align with the new documentation for latency improvements.

prateeksachan commented:
Got it. Will test this out. Any specific type of `ambient-noise` + TTS to choose? Can you share a working agent config?

prateeksachan committed:
Merge branch 'master' into master

abdulrahmanmajid commented:
> Got it. Will test this out. Any specific type of `ambient-noise` + TTS to choose? Can you share a working agent config?

@prateeksachan apologies for the delay, I wasn't alerted for the new reply.

i've added the agent config I used to test but I believe with these changes any TTS should work.


```
{
    "agent_config": {
        "agent_name": "test",
        "agent_type": "other",
        "agent_welcome_message": "Hey!, This is an AI Assistant",
        "tasks": [
            {
                "task_type": "conversation",
                "toolchain": {
                    "execution": "parallel",
                    "pipelines": [
                        [
                            "transcriber",
                            "llm",
                            "synthesizer"
                        ]
                    ]
                },
                "tools_config": {
                    "input": {
                        "format": "wav",
                        "provider": "twilio"
                    },
                    "llm_agent": {
                        "agent_type": "simple_llm_agent",
                        "agent_flow_type": "streaming",
                        "routes": null,
                        "llm_config": {
                            "provider": "openai",
                            "family": "openai",
                            "model": "gpt-4.1-nano",
                            "max_tokens": 50,
                            "temperature": 2,
                            "top_p": 0.5,
                            "presence_penalty": 0,
                            "frequency_penalty": 0
                        }
                    },
                    "output": {
                        "format": "wav",
                        "provider": "twilio"
                    },
                    "synthesizer": {
                        "provider": "azuretts",
                        "stream": true,
                        "buffer_size": 150,
                        "caching": true,
                        "provider_config": {
                            "voice": "AndrewMultilingual",
                            "language": "en-US",
                            "model": "Neural"
                        }
                    },
                    "transcriber": {
                        "provider": "deepgram",
                        "model": "nova-3",
                        "language": "multi",
                        "stream": true,
                        "endpointing": 100
                    }
                },
                "task_config": {
                    "optimize_latency": true,
                    "ambient_noise": true,
                    "ambient_noise_track": "call-center",
                    "incremental_delay": 0,
                    "interruption_backoff_period": 50,
                    "backchanneling": true,
                    "backchanneling_message_gap": "5",
                    "backchanneling_start_delay": "1",
                    "use_fillers": false,
                    "number_of_words_for_interruption": 2,
                    "hangup_after_LLMCall": true
                }
            }
        ]
    },
    "agent_prompts": {
        "task_1": {
            "system_prompt": "you are an AI Assistant"
        }
    }
}
```



with the azure file changes I tried tested the newer azure models, Azure launched DragonHD models which are very nice with context and speech quality but when we use this modal it outputs speech in small bits but i fixed the latneyc issue with this model, in the regular azure file i'd get 2,4 second latency per turn with dragon, but now its 0.2 0.5ms but it is choppy, the other models work just fine with new file.


2nd issue I'm having is either with deepgram or its with twilio, it dosnt listen very well or isnt very accurate when the phone is a desk and on speaker, it only works well when next to ear and I dont think this is a phone issue, the quality of audio that twilio gets is far lower than what we send via tts

prateeksachan committed:
Merge branch 'master' into master

prateeksachan committed:
Merge branch 'master' into master

prateeksachan commented:
Hi @abdulrahmanmajid I have merged this. Wanted to ask:
`get_synthesizer_from_pool` and `return_synthesizer_to_pool`, though defined aren't being used anywhere? Is this code generated via some LLM application?

abdulrahmanmajid commented:
Hello @prateeksachan, Yes it was assisted by cursor AI and Claude 3.7, those were added to create persistent synthesizers to eliminate network latency and connection time, are you sure they aren't being used? I didn't really test it but the tts works great after changes and improved latency by a lot, I also fixed the chunking issue for the larger models like DragonHD, I'll push that soon.
